### DISCUSSION TITLE
[help] Named list names modified in tar_load but not in function call. Is this expected?

### DISCUSSION THREAD

**Discussion by dblodgett-usgs** [ropensci/targets#1117] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Is this treatment of names intended? I've found myself emitting named lists from functions then struggling to work with them interactively / write later functions that work with them because their names have the target prefix appended using `tar_load` but not when loaded directly into a function. 

Sorry if this is documented, I wasn't able to find an answer. Thanks if anyone can help.

I added a few example ways of loading these things at the bottom -- I feel like `tar_load(target, {target index})` should load the target by name and not modify the internal names the same way that loading a data.frame does. Am I crazy?

```r
# _targets.R
library(targets)

test_fun <- function(x)  list(one = x, two = x *2) 

test_fun2 <- function(x) data.frame(one = x, two = x *2)

list(tar_target("input", c(1, 2, 3, 4)),
     tar_target("test", test_fun(input),
                pattern = map(input)),
     tar_target("test2", names(test),
                pattern = map(test)),
     tar_target("test3", test_fun2(input),
                pattern = map(input)))
```

```r
> tar_load(test)
> tar_load(test, 1)
> tar_make()
#omitted output
 
> tar_load(test)
> constructive::construct(test)
list(
  test_5811020d_one = 1,
  test_5811020d_two = 2,
  test_4143de2f_one = 2,
  test_4143de2f_two = 4,
  test_354b50fa_one = 3,
  test_354b50fa_two = 6,
  test_f0316d7e_one = 4,
  test_f0316d7e_two = 8
)

> tar_load(test2)
> constructive::construct(test2)
c(
  test2_0c4e4cfd_1 = "one", test2_0c4e4cfd_2 = "two", test2_802200b3_1 = "one",
  test2_802200b3_2 = "two", test2_92fdb7d3_1 = "one", test2_92fdb7d3_2 = "two",
  test2_026c6b93_1 = "one", test2_026c6b93_2 = "two"
)

> tar_read(test, 1)
$test_5811020d_one
[1] 1

$test_5811020d_two
[1] 2

> tar_read(test_5811020d)
$one
[1] 1

$two
[1] 2

tar_load(tar_branch_names("test", 1))
# loaded test_5811020d
> constructive::construct(test_5811020d)
list(one = 1, two = 2)

# but if we have a set of data.frames, we don't get any record of which row a branch created and names are clean
> tar_load(test3)
> constructive::construct(test3)
data.frame(one = c(1, 2, 3, 4), two = c(2, 4, 6, 8))
```

**Comment by wlandau** [comment, marked as answer]
This happens because the default behavior is to treat dynamic targets as "vectors" (the generalization of a vector from the `vctrs` package, which includes data frames) with branches as elements. With that setup, if a branch has more than one component, then all components together are flattened and given unique names. The setting that controls this is `iteration = "vector"` in `tar_target()` and `tar_option_set()`. For `test1` and `test2`, you will have better luck with `iteration = "list"`.

**Reply by dblodgett-usgs** [reply]
🙏
